### PR TITLE
added task solution

### PR DESCRIPTION
**Pull request by urban1991**

### CODE REVIEW

File: src/averageAges.js [hunk 1]
```diff
@@ -37,7 +42,16 @@ function calculateMenAverageAge(people, century) {
...
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = withChildren
+    ? people.filter((p) => p.sex === 'f'
+      && people.find((p1) => p1.mother === p.name))
+    : people.filter((p) => p.sex === 'f');
+
+  let sum = 0;
+
+  women.forEach((w) => sum += w.died - w.born);
```
Comment (urban1991) [inline src/averageAges.js:52]: Similar as above, if you must return one value the reduce method is way better for that task.

File: src/averageAges.js [hunk 2]
```diff
@@ -37,7 +42,16 @@ function calculateMenAverageAge(people, century) {
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = withChildren
+    ? people.filter((p) => p.sex === 'f'
+      && people.find((p1) => p1.mother === p.name))
+    : people.filter((p) => p.sex === 'f');
+
+  let sum = 0;
```
Comment (urban1991) [inline src/averageAges.js:50]: Redundant if you use `.reduce`

File: src/averageAges.js [hunk 3]
```diff
@@ -37,7 +42,16 @@ function calculateMenAverageAge(people, century) {
...
+
+  women.forEach((w) => sum += w.died - w.born);
+
+  return sum / women.length;
```
Comment (urban1991) [inline src/averageAges.js:54]: Missing error handlers

File: src/averageAges.js [hunk 4]
```diff
@@ -37,7 +42,16 @@ function calculateMenAverageAge(people, century) {
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = withChildren
+    ? people.filter((p) => p.sex === 'f'
+      && people.find((p1) => p1.mother === p.name))
+    : people.filter((p) => p.sex === 'f');
```
Comment (urban1991) [inline src/averageAges.js:48]: Currently, this fn don't handle the edge case where there are no matching women (which would result in a division by zero when calculating the average age).

File: src/averageAges.js [hunk 5]
```diff
@@ -37,7 +42,16 @@ function calculateMenAverageAge(people, century) {
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = withChildren
+    ? people.filter((p) => p.sex === 'f'
```
Comment (urban1991) [inline src/averageAges.js:46]: As above this code is complicated. Try one liner using `.filter` and  `.some` methods

File: src/averageAges.js [hunk 6]
```diff
@@ -55,7 +69,19 @@ function calculateWomenAverageAge(people, withChildren) {
...
-  // write code here
+  const children = onlyWithSon
+    ? people.filter(p => p.sex === 'm'
+      && people.some(m => m.name === p.mother))
+    : people.filter(p => people.some(m => m.name
+      === p.mother));
+
+  let sum = 0;
+
+  children.forEach((ch) => sum += ch.born - people.find(
```
Comment (urban1991) [inline src/averageAges.js:80]: Same as above, `.reduce` is better choice.

File: src/averageAges.js [hunk 7]
```diff
@@ -55,7 +69,19 @@ function calculateWomenAverageAge(people, withChildren) {
...
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const children = onlyWithSon
+    ? people.filter(p => p.sex === 'm'
+      && people.some(m => m.name === p.mother))
+    : people.filter(p => people.some(m => m.name
+      === p.mother));
+
+  let sum = 0;
```
Comment (urban1991) [inline src/averageAges.js:78]: To be dropped when using `.reduce`

File: src/averageAges.js [hunk 8]
```diff
@@ -55,7 +69,19 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const children = onlyWithSon
+    ? people.filter(p => p.sex === 'm'
```
Comment (urban1991) [inline src/averageAges.js:73]: Try to simplify this condition using the same approach as i mention above in `calculateWomenAverageAge` function

File: src/averageAges.js [hunk 9]
```diff
@@ -55,7 +69,19 @@ function calculateWomenAverageAge(people, withChildren) {
...
+    : people.filter(p => people.some(m => m.name
+      === p.mother));
+
+  let sum = 0;
+
+  children.forEach((ch) => sum += ch.born - people.find(
+    (m) => m.name === ch.mother
+  ).born);
+
+  return sum / children.length;
```
Comment (urban1991) [inline src/averageAges.js:84]: Add error handlers

**Review by urban1991 — COMMENTED** [review]
In conclusion, these functions are a good start and implement the necessary logic to solve the given problems. However, they could be improved by applying more functional programming principles (like using reduce instead of forEach) and adding error handling to deal with edge cases. As always, it's important to thoroughly test these functions with different inputs to ensure they work as expected in all scenarios.

File: src/averageAges.js (hunk 8 above)
Comment (urban1991) [inline src/averageAges.js:73]: This condition seems to be appearing quite often. If such a situation arises, consider creating a helper function to prevent repetition.
